NLB-6626: Add documentation for dataplane managed identity (#1748)
nginx/documentation: content/nginxaas-azure/quickstart/dataplane-mi.md

---
title: NGINXaaS Managed Identity on dataplane
weight: 700
toc: true
url: /nginxaas/azure/quickstart/dataplane-mi/
type:
- how-to
---

## Overview

> F5 NGINX as a Service for Azure (NGINXaaS) supports using the Managed Identity(MI) assigned to the NGINXaaS deployment to access other Azure resources in the same virtual network or those that are accessable publically.

Reviewer: ```suggestion
F5 NGINX as a Service for Azure (NGINXaaS) supports using the Managed Identity (MI) assigned to the NGINXaaS deployment to access other Azure resources in the same virtual network or those that are publicly accessible.
```

    } catch (err) {
        r.error(`ERROR: Exception occurred while querying IMDS. Details: ${err}`);
        return null;
    }
}

async function queryBlob(r) {
    const storageAccountName = "test-storage";
    const containerName = "test-container";
    const apiEndpoint = `https://${storageAccountName}.blob.core.windows.net/${containerName}?restype=container&comp=list`;

    r.log("fetching access token...");
    const accessToken = await fetchAccessToken(r);

    if (!accessToken) {
        r.return(401, "Failed to fetch access token.");
        return;
    }

    r.log(`INFO: querying blob storage... ${apiEndpoint}`);
    try {
        const reply = await ngx.fetch(apiEndpoint, {
            method: "GET",
            headers: {
                "Authorization": `Bearer ${accessToken}`,
                "x-ms-version": "2019-12-12"
            },
            verify: false
        });

        const body = await reply.text();
        r.return(reply.status, body);
    } catch (err) {
        r.error(`ERROR: Exception occurred while querying Blob Storage. Details: ${err}`);
        r.return(500, "Failed to query Blob Storage.");
    }
}

export default { queryBlob };

> ```

Reviewer: ````suggestion
    ```javascript
    async function fetchAccessToken(r) {
        const resource = "https://storage.azure.com/";
        const apiVersion = "2019-08-01";
        const imdsEndpoint = `http://169.254.169.254/metadata/identity/oauth2/token?resource=${resource}&api-version=${apiVersion}`;
        try {
            const imdsResponse = await ngx.fetch(imdsEndpoint, {
                headers: { "Metadata": "true" }
            });

            if (imdsResponse.status !== 200) {
                let resp = JSON.stringify(imdsResponse);
                r.error(`ERROR: Failed to fetch access token. IMDS returned status: ${resp}`);
                return null;
            }
            let body = await imdsResponse.text();
            const imdsResponseBody = JSON.parse(body);

            let token = imdsResponseBody.access_token;
            return token;
        } catch (err) {
            r.error(`ERROR: Exception occurred while querying IMDS. Details: ${err}`);
            return null;
        }
    }

    async function queryBlob(r) {
        const storageAccountName = "test-storage";
        const containerName = "test-container";
        const apiEndpoint = `https://${storageAccountName}.blob.core.windows.net/${containerName}?restype=container&comp=list`;

        r.log("fetching access token...");
        const accessToken = await fetchAccessToken(r);

        if (!accessToken) {
            r.return(401, "Failed to fetch access token.");
            return;
        }

        r.log(`INFO: querying blob storage... ${apiEndpoint}`);
        try {
            const reply = await ngx.fetch(apiEndpoint, {
                method: "GET",
                headers: {
                    "Authorization": `Bearer ${accessToken}`,
                    "x-ms-version": "2019-12-12"
                },
                verify: false
            });

            const body = await reply.text();
            r.return(reply.status, body);
        } catch (err) {
            r.error(`ERROR: Exception occurred while querying Blob Storage. Details: ${err}`);
            r.return(500, "Failed to query Blob Storage.");
        }
    }

    export default { queryBlob };
    ```
````

indent the code block under the section

Sending a http request to the queryBlob endpoint will invoke njs which will query IMDS to get access tokens and use the tokens to query blob storage.

> {{<note>}} IMDS enforces a rate limit of 5 requests per second. To optimize performance, cache the access token in NGINX instead of retrieving it for every request.{{</note>}}

Reviewer: ```suggestion
{{< call-out "note" >}} IMDS enforces a rate limit of 5 requests per second. To optimize performance, cache the access token in NGINX instead of retrieving it for every request.{{</ call-out "note" >}}
```
